### PR TITLE
tools/mpy-tool.py: Force function alignment to 4 bytes.

### PR DESCRIPTION
**Pull request by jimmo**
This appears to be necessary for thumb. Without this change, calling a frozen native/viper function will crash.

### CODE REVIEW

**Comment by dpgeorge** [comment]
I thought it would be aligned by virtue of being in the `.text` section, but perhaps not.

It'd be good to try and test this issue with CI somehow, possibly with frozen native code in ports/qemu-arm.

**Comment by jimmo** [comment]
The type is `byte[]`, so it will be single-byte aligned by default.

Here's one I caught in the wild:

before this change:
```
$ nm build-PYBV11/firmware.elf  | grep fun_data_native_test
0803ed4b t fun_data_native_misc__lt_module_gt__native_test
```

after:
```
$ nm build-PYBV11/firmware.elf  | grep fun_data_native_test
0803ed50 t fun_data_native_misc__lt_module_gt__native_test
```

I agree it would be good to get this in CI... any suggestions on how to test this specific issue though. (i.e. how to write a .py file such that the generated frozen file will always contain a misaligned function). I was able to repro the failure in the qemu-arm build though (and qemu fails):

```
$ nm ports/qemu-arm/build/firmware.elf | grep fun_data_native_
00024fc0 t fun_data_native_test__lt_module_gt_
0001a945 t fun_data_native_test__lt_module_gt__native_test
0001a9d8 t fun_data_native_test__lt_module_gt__native_x
0001a90c t fun_data_native_test__lt_module_gt__native_y
```

In general though, CI (QEMU and Unix) for frozen modules (with and without native code) would be good. How do you feel about adding ports/qemu-arm/test-modules/*.py and enabling that for Makefile.test, and add some tests (skipped on other platforms) that exercise them. (Maybe in a different PR though?)

**Comment by dpgeorge** [comment]
> ```
> 0803ed4b t fun_data_native_misc__lt_module_gt__native_test
> ```

Ok, so it's definitely needed.

But it should only need to be 2-byte aligned, on ARMxxM targets.  ARMV6 would probably need 4 bytes alignment, and  x86 and xtensa should only need 1 byte alignment (so nothing to add there).

----

> any suggestions on how to test this specific issue though.

It's hard.... probably ok to just have a decent number of frozen native functions, hoping that it would cover enough cases to catch this error.

> In general though, CI (QEMU and Unix) for frozen modules (with and without native code) would be good.

unix coverage already has frozen modules, see `ports/unix/coverage-frzmpy`.  So more tests could just be added to that dir, for unix.  For qemu-arm, a simple solution would be to just reuse this unix coverage dir, but that could get a bit messy.

Alternatively one could freeze the entire test suite with native code generation enabled (or better, just a subdir like `tests/basics`), but that would lead to very big firmware (qemu-arm is limited to 1 or 2 mbytes) and probably other issues.

**Comment by peterhinch** [comment]
This works well here.

**Comment by jimmo** [comment]
> But it should only need to be 2-byte aligned, on ARMxxM targets. ARMV6 would probably need 4 bytes alignment, and x86 and xtensa should only need 1 byte alignment (so nothing to add there).

OK, I have made it conditional on the value of `native_arch`.

> Alternatively one could freeze the entire test suite with native code generation enabled (or better, just a subdir like tests/basics), but that would lead to very big firmware (qemu-arm is limited to 1 or 2 mbytes) and probably other issues.

I am keen to do this, even at the very least just a subset (e.g. tests/micropython), but for now I've just done the basic plumbing to make this possible later, and added a single test that reproduces this issue. (Without the alignment change it now fails with a timeout, otherwise it passes)

File: ports/qemu-arm/Makefile [hunk 1]
```diff
@@ -1,6 +1,8 @@
 include ../../py/mkenv.mk
 -include mpconfigport.mk
 
+FROZEN_MPY_DIR ?= modules
```
Comment (dpgeorge) [inline ports/qemu-arm/Makefile]: Will standard (non-test) qemu-arm build with this, given the directory doesn't exist?  Probably this line should be deleted.

File: ports/qemu-arm/Makefile.test [hunk 2]
```diff
@@ -19,6 +21,7 @@ $(BUILD)/firmware-test.elf: $(OBJ_COMMON) $(OBJ_TEST)
 	$(Q)$(SIZE) $@
 
 test: $(BUILD)/firmware-test.elf
-	qemu-system-arm -machine $(BOARD) $(QEMU_EXTRA) -nographic -monitor null -semihosting -kernel $< > $(BUILD)/console.out
+    # qemu hangs (e.g. due to alignment errors).
```
Comment (dpgeorge) [inline ports/qemu-arm/Makefile.test]: Please put comment just above the rule, else it'll be printed as part of the execution of the rule (unless that's what's intended...).  Also maybe something like "timeout is to stop qemu if it hangs"

File: tests/run-tests [hunk 3]
```diff
@@ -537,24 +538,28 @@ the last matching regex is used:
...
-                    'basics', 'micropython', 'float', 'import', 'io', 'misc',
-                    'stress', 'unicode', 'extmod', 'unix', 'cmdline',
-                )
+                test_dirs += ('float', 'import', 'io', 'stress', 'unicode', 'unix', 'cmdline',)
+            elif args.target == 'qemu-arm':
+                if not args.write_exp:
+                    raise ValueError('--target=qemu-arm must be used with --write-exp')
+                # Generate expected output files for qemu run.
+                # This list should match the test_dirs tuple in tinytest-codegen.py.
+                test_dirs += ('float', 'inlineasm', 'qemu-arm',)
```
Comment (dpgeorge) [inline tests/run-tests]: It looks like the lists don't match because tinytest-codegen.py doesn't include 'misc'.  Maybe misc can be enabled on qemu-arm?

**Comment by jimmo** [comment]
> Maybe misc can be enabled on qemu-arm?

Yes, with one exception (`misc/print_exception.py`) which I have added to the skip list for qemu-arm.

Other comments have been addressed.

**Comment by dpgeorge** [comment]
Thanks!  Merged in 4ab5156c01bfd4a6304e26b1dc2d34163b8637c0 and 0bd1eb80ff49dd0d4ad0c369a83a5aadea995944

**Pull request closed without merging**